File: cpp/Karatsubamultiplication.cpp
// karatsuba_multiplication.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <chrono>
#include <vector>

using namespace std;
using Clock = chrono::high_resolution_clock;

// Remove leading zeros from a numeric string
string trimZeros(const string &s) {
    string r = s;
    while (r.size() > 1 && r[0] == '0') r.erase(r.begin());
    return r;
}

// Add two numbers represented as strings
string addStrings(const string &a, const string &b) {
    string res = "";
    int carry = 0, i = a.size() - 1, j = b.size() - 1;

    while (i >= 0 || j >= 0 || carry) {
        int sum = carry;
        if (i >= 0) sum += a[i--] - '0';
        if (j >= 0) sum += b[j--] - '0';
        carry = sum / 10;
        res.push_back(char((sum % 10) + '0'));
    }
    reverse(res.begin(), res.end());
    return trimZeros(res);
}

// Subtract b from a (a >= b guaranteed)
string subtractStrings(const string &a, const string &b) {
    string res = "";
    int borrow = 0, i = a.size() - 1, j = b.size() - 1;

    while (i >= 0) {
        int diff = (a[i] - '0') - borrow - (j >= 0 ? (b[j] - '0') : 0);
        if (diff < 0) diff += 10, borrow = 1;
        else borrow = 0;
        res.push_back(char(diff + '0'));
        i--; j--;
    }

    reverse(res.begin(), res.end());
    return trimZeros(res);
}

// Multiply number by 10^n (append zeros)
string shiftLeft(const string &s, int n) {
    return s + string(n, '0');
}

// Karatsuba recursive multiplication
string karatsuba(const string &x, const string &y) {
    string a = trimZeros(x);
    string b = trimZeros(y);

    // Base case for small numbers (direct multiplication)
    if (a.size() < 10 && b.size() < 10) {
        long long n1 = stoll(a);
        long long n2 = stoll(b);
        return to_string(n1 * n2);
    }

    int n = max(a.size(), b.size());
    if (n % 2) n++; // make even length

    // Pad with leading zeros
    while ((int)a.size() < n) a = "0" + a;
    while ((int)b.size() < n) b = "0" + b;

    int half = n / 2;

    // Split into halves
    string aL = a.substr(0, n - half);
    string aR = a.substr(n - half);
    string bL = b.substr(0, n - half);
    string bR = b.substr(n - half);

    // Karatsuba recursive calls
    string p1 = karatsuba(aL, bL);
    string p2 = karatsuba(aR, bR);
    string p3 = karatsuba(addStrings(aL, aR), addStrings(bL, bR));

    // Combine results:
    // result = p1 * 10^(2*half) + (p3 - p1 - p2) * 10^(half) + p2
    string mid = subtractStrings(subtractStrings(p3, p1), p2);

    string result = addStrings(
        addStrings(shiftLeft(p1, 2 * half), shiftLeft(mid, half)),
        p2
    );

    return trimZeros(result);
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    cout << "Karatsuba Multiplication Benchmark\n";
    cout << "Enter first large integer: ";
    string x;
    cin >> x;

    cout << "Enter second large integer: ";
    string y;
    cin >> y;

    // Start timing
    auto start = Clock::now();

    string result = karatsuba(x, y);

    // End timing
    auto end = Clock::now();
    chrono::duration<double> elapsed = end - start;

    cout << "\nProduct:\n" << result << "\n";
    cout << "\nExecution time: " << elapsed.count() << " seconds\n";

    return 0;
}

File: Python/KaratsubaMultiplication.py
import time

# Remove leading zeros
def trim(s):
    s = s.lstrip("0")
    return s if s != "" else "0"

# Add two numeric strings
def add_str(a, b):
    a = a[::-1]
    b = b[::-1]
    carry = 0
    result = []

    for i in range(max(len(a), len(b))):
        da = int(a[i]) if i < len(a) else 0
        db = int(b[i]) if i < len(b) else 0
        s = da + db + carry
        result.append(str(s % 10))
        carry = s // 10

    if carry:
        result.append(str(carry))

    return trim("".join(result[::-1]))

# Subtract b from a (assuming a >= b)
def sub_str(a, b):
    a = a[::-1]
    b = b[::-1]
    result = []
    borrow = 0

    for i in range(len(a)):
        da = int(a[i])
        db = int(b[i]) if i < len(b) else 0
        d = da - db - borrow
        if d < 0:
            d += 10
            borrow = 1
        else:
            borrow = 0
        result.append(str(d))

    return trim("".join(result[::-1]))

# Multiply by 10^n (just append n zeros)
def shift(s, n):
    return s + "0" * n

# Standard integer multiplication for small values
def small_mul(a, b):
    return str(int(a) * int(b))

# Karatsuba multiplication
def karatsuba(a, b):
    a = trim(a)
    b = trim(b)

    # Base case: small enough for direct multiply
    if len(a) < 10 and len(b) < 10:
        return small_mul(a, b)

    # Normalize length to be equal and even
    n = max(len(a), len(b))
    if n % 2 == 1:
        n += 1

    a = a.zfill(n)
    b = b.zfill(n)
    half = n // 2

    aL, aR = a[:n-half], a[n-half:]
    bL, bR = b[:n-half], b[n-half:]

    # Recursive calls
    p1 = karatsuba(aL, bL)
    p2 = karatsuba(aR, bR)
    p3 = karatsuba(add_str(aL, aR), add_str(bL, bR))

    # Combine Karatsuba results:
    # result = p1*10^(2*half) + (p3 - p1 - p2)*10^(half) + p2
    mid = sub_str(sub_str(p3, p1), p2)

    result = add_str(
        add_str(shift(p1, 2 * half), shift(mid, half)),
        p2
    )
    return trim(result)

# ---------------------------- MAIN PROGRAM ----------------------------
if __name__ == "__main__":
    print("Karatsuba Multiplication Benchmark")

    x = input("Enter first large integer: ").strip()
    y = input("Enter second large integer: ").strip()

    start = time.perf_counter()
    result = karatsuba(x, y)
    end = time.perf_counter()

    print("\nProduct:")
    print(result)
    
    print("\nExecution time:", end - start, "seconds")
